pkg/router/router.go

```diff
@@ -302,6 +303,14 @@ func registerPropertiesEndpoint(r endpoints.Resource, container *restful.Contain
 	container.Add(wsDefaults)
 }
 
+func registerCSRFTokenEndpoint(r endpoints.Resource, container *restful.Container) {
+	ws := new(restful.WebService)
+	ws.Filter(restful.NoBrowserCacheFilter)
+	ws.Path("/v1/token").Produces("text/plain")
+	ws.Route(ws.GET("/").To(r.GetToken))
```

Comment:
Question here, and this is more so I can understand it fully - on anything but OpenShift where we've got the secure proxy on top (let's say they're on GKE or AWS), what's to stop the malicious party from visiting this endpoint first to get a token? It'd be pointless though wouldn't it because it's a new token on top of every single request we do that can alter state?

Reply:
There are 2 main pieces to this, a cookie and a custom header, both must be provided to pass the check.

CSRF relies on the default browser behaviour of forwarding cookies on requests to a given domain. So an attacker would craft a link / form on a page they control, that navigates / submits to the target domain. If a user visits this malicious page and follows the link (or the form is auto-submitted using a script on the page) the browser will happily forward any cookies it has for the target. If the user is already authenticated to that system, the attacker now has the ability to perform actions in the system as that user.

To mitigate against this:
- do not use GET for mutations, this prevents malicious links from causing harm
- set SameSite=Lax on session / auth cookies (this is the default in most modern browsers when SameSite is omitted), this means that in a cross-domain scenario they'll only be submitted on GET requests which should be safe (point above)
- since we may not be able to set the SameSite attribute on all affected cookies (may not control them ourselves) we can take additional steps:
  - require a custom header on requests containing some random token that can be matched against some server-side / session value
  - the token value cannot be retrieved by scripts on another origin (same origin protection on reading response body / headers)

gorilla/csrf uses the 'double submit' method where masked tokens are generated for each request that must be submitted in a custom header as well as a cookie that will be automatically provided by the browser on subsequent requests. For any 'unsafe' methods (POST, PUT, DELETE, etc.) these values are unmasked on the server and compared against the current key to verify them.

Summary:
- need to provide both cookie + header
- cannot read cookies cross-origin
- cannot read response body / headers cross-origin
- by default modern browsers will not submit cookies on 'unsafe' requests made cross-origin

Reply:
Thanks Alan, excellent explanation 👍